# [Genome540](http://bozeman.mbt.washington.edu/compbio/mbt599/ "Course Page")
## Introduction to Computational Molecular Biology: Genome and Protein Sequence Analysis

Genome 540 is the first quarter of a two-quarter introduction to protein and DNA sequence analysis and molecular evolution, including probabilistic models of sequences and of sequence evolution, computational gene identification, pairwise sequence comparison and alignment (algorithms and statistical issues), multiple sequence alignment and evolutionary tree construction, comparative genomics, and protein sequence/structure relationships. These represent the central computational approaches for interpreting genomes. The statistical and algorithmic ideas discussed (which include maximum likelihood estimation, hidden Markov models, dynamic programming) have wide applicability in other areas of computational & mathematical biology as well.

Homework 1: Suffix Arrays, UltraConserved Elements (UCE)

Homework 2: Order-0/Order-1 Markov Models

Homework 3: Log-Likelihood Ratios (LLR), CoDing Sequences (CDS)

Homework 4: Dynamic Programming, Weighted Directed Acyclic Graphs (WDAG)

Homework 5: Topological Sort, Network Optimization, Local Sequence Alignment, BLOSUM62

Homework 6: D-segment Algorithm, Poisson Distributions

Homework 7: Next-Generation Sequencing (NGS), Karlin-Altschul Theory

Homework 8: Hidden Markov Models (HMM), Parameter Estimation (EM/Baum-Welch Training)

Homework 9: Viterbi Algorithm, Evolutionarily Conserved Segments (ECS), Putative Neutral/Functional Sequences


# Genome541
## Introduction to Computational Molecular Biology: Molecular Evolution

Genome 541 provides a survey of topics within the field of computational molecular biology. The course is divided into five two-week blocks, each devoted to a single topic and taught by a different instructor. This year, the topics include protein structure, single-cell analysis, epigenomics, cancer genomics, and phylogenetics.

<ins>Protein Structure</ins>

Homework 1: Protein Energetics/Modeling, Ramachandran Space (phi-psi torsion angles)

Homework 2: Simulated Annealing Monte Carlo, Verlet Algorithm, Mfold Algorithm, Fragment Folding

<ins>Single-Cell Analysis</ins>

Homework 3: Manifold Learning, KD-Tree/Ball Tree, (Absorbing) Markov Chains, Diffusion Maps, Trajectory Inference, RNA Velocity

Homework 4: Variational Autoencoders (VAE), Gene Regulatory Networks, Cross-Modality Prediction

